rules/endpoints.go

```diff
@@ -403,16 +414,31 @@ func (r *DefaultRuleRenderer) endpointIptablesChain(
 				})
 		}
 
-		// When rendering normal rules, if no profile marked the packet as accepted, drop
-		// the packet.
+		// When rendering normal rules, except for the wildcard HEP, drop the packet if no
+		// policy or profile allowed it.
 		//
-		// For untracked rules, we don't do that because there may be tracked rules
-		// still to be applied to the packet in the filter table.
-		rules = append(rules, Rule{
-			Match:   Match(),
-			Action:  DropAction{},
-			Comment: []string{"Drop if no profiles matched"},
-		})
+		// For untracked and pre-DNAT rules, we don't do that because there may be tracked
+		// rules still to be applied to the packet in the filter table.
+		//
+		// For the wildcard HEP, we allow when there is no policy, because wildcard HEPs
+		// were previously implemented only for pre-DNAT policy and so had no interaction
+		// with normal policy.
+		if defaultDropWhenNoPolicy {
```

Comment:
Traffic will fall through to these rules if there is policy but it doesn't match.  Is that what we want?  I was expecting that we'd only want to do default-allow if there is no policy at all.

Reply:
(But please check my expectation since this epic has been pretty fast-moving so I may be out-of-date!)

Reply:
I think your expectation is right.  But I think your interpretation of what the code does is wrong, because of "Drop if no policies passed packet" just above (https://github.com/projectcalico/felix/pull/2228/files#diff-0370ffa1b8a68941ac13794784606c0aL374).

Reply:
(It's a separate point, but note that we rather awkwardly have two different approaches for the no-policy-at-all case.  For some types of policy it's handled by not calling into the endpoint chain at all.  For others it's handled by code like here and at https://github.com/projectcalico/felix/pull/2228/files#diff-0370ffa1b8a68941ac13794784606c0aL378.)

Reply:
((To be clear, that awkwardness ^ was not introduced by this PR.))

Reply:
OK, I see that rule now; but what about profiles?  Should we apply the same logic there?  I.e. if there are some profiles then should that imply default drop?

Reply:
> OK, I see that rule now; but what about profiles? Should we apply the same logic there? I.e. if there are some profiles then should that imply default drop?

It seems more consistent and I think this still works with pre-dnat policy (at the very least, the existing pre-dnat tests pass). @neiljerram wdyt?

Reply:
E.g., it seems that in addition to having to provide egress normal policy rules like we did https://github.com/projectcalico/felix/pull/2228/commits/7d84dd5fe8bd61d1e1455e9342750a2d2b673cd6, if we were to drop if no profiles applied as proposed, we'd also need to provide ingress policy rules on the destination host.

Reply:
I suppose what I'm getting at is that there's a difference between:

> there are no profiles

and

> there are profiles but none of them matched the packet

I think this treats those equally, whereas, for policies, we have a default drop if there are policies but none of the match.

Reply:
IIUC, do you mean a change like this?
```
@@ -440,7 +440,7 @@ func (r *DefaultRuleRenderer) endpointIptablesChain(
                // with normal policy.
                if options.defaultDropWhenNoPolicy {
                        rules = append(rules, Rule{
-                               Match:   Match(),
+                               Match:   Match().MarkClear(r.IptablesMarkPass),
                                Action:  DropAction{},
                                Comment: []string{"Drop if no profiles matched"},
                        })
```

master doesn't make this distinction either: https://github.com/projectcalico/felix/blob/master/rules/endpoints.go#L411-L414

Reply:
> master doesn't make this distinction either

Right, master drops in both cases.  Now you allow in both cases for a WHEP.  But for policies, we allow in one case and drop in the other.  I'm wondering if that's right, or if we should allow in the "no profiles" case and drop in the "some profiles but none allowed the packet" case

Reply:
We resolved this in Slack. WHEPs will behave similarly to regular named HEPs regarding profiles.